trans/train.py

```diff
@@ -147,30 +200,28 @@ def main(args: argparse.Namespace):
 
         logging.info("Training...")
         transducer_.train()
+        transducer_.zero_grad()
         with utils.Timer():
             train_loss = 0.
-            rollin = rollin_schedule(epoch)
+            # rollin not implemented at the moment
+            # rollin = rollin_schedule(epoch)
             j = 0
             for j, batch in enumerate(training_data_loader):
-                losses = []
-                transducer_.zero_grad()
-                for sample in batch:
-                    output = transducer_.transduce(
-                        input_=sample.input,
-                        encoded_input=sample.encoded_input,
-                        target=sample.target,
-                        rollin=rollin,
-                    )
-                    # split losses --> optimize for all optimal actions
-                    losses.extend([s for loss in output.losses for s in torch.split(loss, 1)])
-                batch_loss = -torch.mean(torch.stack(losses))
-                train_loss += batch_loss.item()
-                batch_loss.backward()
-                optimizer.step()
+                losses = transducer_.training_step(encoded_input=batch.encoded_input,
+                                                   action_history=batch.action_history,
+                                                   alignment_history=batch.alignment_history,
+                                                   optimal_actions_mask=batch.optimal_actions_mask,
+                                                   valid_actions_mask=batch.valid_actions_mask)
+                train_loss += torch.mean(losses.squeeze(dim=0)).item()  # mean per batch
+                losses.sum().backward()
+                if j % args.grad_accumulation == 0:
+                    optimizer.step()
+                    transducer_.zero_grad()
                 if j > 0 and j % 100 == 0:
                     logging.info("\t\t...%d batches", j)
             logging.info("\t\t...%d batches", j + 1)
 
+        # avg. loss per sample
         avg_loss = train_loss / len(training_data_loader)
```

Comment:
Shouldn't `train_loss` be divided by the the number of batches rather than the number of samples?

Reply:
Yes, `len(training_data_loader)` is equal to the number of batches.